gobblin-runtime/src/main/java/gobblin/runtime/cli/CliEmbeddedGobblin.java

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.runtime.cli;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+import org.apache.commons.cli.CommandLine;
+import org.apache.commons.cli.CommandLineParser;
+import org.apache.commons.cli.DefaultParser;
+import org.apache.commons.cli.HelpFormatter;
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.Options;
+import org.apache.commons.cli.ParseException;
+
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+
+import gobblin.annotation.Alias;
+import gobblin.runtime.embedded.EmbeddedGobblin;
+import gobblin.util.ClassAliasResolver;
+
+import lombok.Data;
+
+
+/**
+ * A {@link CliApplication} that runs a Gobblin flow using {@link EmbeddedGobblin}.
+ *
+ * Usage:
+ * java -jar gobblin.jar run [app] [OPTIONS]
+ *
+ * The cli can run arbitrary applications by either setting each configuration manually (using the -setConfiguration
+ * option), or with the help of a template (using the -setTemplate option). However, since this requires knowing
+ * the exact configurations needed for a job, developers can implement more specific applications with simpler options
+ * for the user.
+ *
+ * {@link CliEmbeddedGobblin} will load any extension of {@link EmbeddedGobblin} using an alias or a class name. It will
```

Comment:
This looks a bit too restrictive on what the input parameters can be. For example, it will be hard to model a standard "cp source1 ... sourceN dest" command line. Why not let the application/tool take control of the command line? For example, we can have a factory instead of annotation

interface CliAccess {
   Options getOptions();
   void run(CommandLine cli) throws CliParseException, CliRunException;
}

Applications can provide an implementation of the interface that can take full use of the apache commons parsing capabilities like long/short names, argument names, multiple arguments, required parameters with option names, etc. You can move some of the annotation magic to a helper class that CliAccess implementations can use to create the command-line options list.